### DISCUSSION TITLE
Spawn trains evenly between stations on multi-station rides

### DISCUSSION THREAD

**Discussion by ocalhoun6** [OpenRCT2/OpenRCT2#17842] [Feature requests] [answered]
Currently, when you open/test a ride for the first time that has multiple stations and multiple trains, all trains will spawn at station 1.

I think it would be much better (and theoretically not _that_ difficult to accomplish) to instead evenly divide the number of trains by the number of stations and spawn that number of trains at each station.

For example:
- a mobius coaster with 2 stations and 2 trains would spawn 1 train at each station
- a mobius coaster with 2 stations and 4 trains would spawn 2 trains at each station
- a monorail with 4 stations and 12 trains would spawn 3 trains at each station
- a log flume with 3 stations and 20 (single-car) trains would spawn 7 trains at station 1, 6 trains at station 2, and 6 trains at station 3
- a car ride with 1 entrance station and 1 exit station and 13 trains would spawn 7 trains at station 1 and 6 trains at station 2 (and then, of course, the trains at the exit-only station would proceed to the next station as soon as allowed)

Advantages:
- Synchronized mobius coasters would be much easier to start, and to restart after a crash or something. Currently, restarting a mobius coaster means you have to manually deselect 'synchronize stations', let half the trains depart station 1, and then click synchronize stations again ... and the ride can't really start processing guests through until the first half of the trains have reached station 2. That's kind of a pain to wait for, and it's inconvenient. And if there are several trains, it's easy to mess up and let too many/too few trains pass before enabling synchronization. But with this feature, mobius coasters could be started immediately, without needing to change the ride settings.
- Long, multi-station transport rides will more quickly and more efficiently spread trains around the track. Currently, all the trains start bunched up at station 1, and for a long transport ride, it might take quite a while for the trains to disperse more or less evenly through the rest of the track. But if they spawn evenly spaced out between stations, that would go a long way toward getting the trains more spread out from the beginning.
- Long, multi-station rides that are slow (like a big, multi-station splash boats) will benefit from spawning trains at all stations. Currently, you often have a situation where the queue for station 1 is empty with lots of boats waiting at station 1 for guests, while the queues for all the other stations are packed full of guests, waiting for the boats from station 1 to finally get there. Splitting boats evenly between the stations would help make wait times at the different stations more similar to each other from the very beginning, helping the ride run more efficiently.
- I think it's a bit more realistic. In the situations listed above, a real-life park would definitely opt for the more efficient way and start their trains simultaneously at multiple stations when possible.

Disadvantages:
- There might be a potential for bugs when doing this with rides that can have more trains than fit in the station (like the miniature railroad or car ride), particularly if there are two stations very close together. Say, if station 2 is only two tiles behind station 1, then the overflow from station 1 might spawn trains within station 2 ... and then station 2 tries to spawn _more_ trains right on top of them. Hopefully, station 1's trains would pass right through station 2's trains until they cleared the space and made room for station 2's trains to move, but even in that best case, there would still be visual issues with the trains overlapping. Or there might be even worse glitches/bugs when the game tries to have two trains in the same spot, or gets train 5 stuck behind train 10, etc. A potential fix for this may be to have station 1 spawn trains first, then station 2, etc ... and when each successive station spawns trains, they check the first tile of the station to see if there's already a train there. If there is, they abort spawning trains and tell station 1 to spawn more trains instead.
- Another potential source of bugs is if you have some kind of hacked ride or plugin or something and find a way to open/test a ride that has no stations at all. Then when it tries to divide the number of trains by the number of stations, there's potential for getting a divide by zero error and a crash. But that should be easy enough to guard against.

**Comment by kyphii** [comment]
This got me thinking so I whipped up a proof of concept:

![stationes](https://user-images.githubusercontent.com/8711258/218875917-d657bf87-51a3-42bd-9056-233b9fa14912.gif)

This touches a lot of the code surrounding Ride::CreateVehicles(). It first gets a list of all the valid spawn locations (stations, and if block sections are used, block brakes and chain lift tops). When spawning the trains, it divides up the trains per station as evenly as possible - any trains left over go to the last station.

This introduces some potential for improvements over the current implementation for block sections. Currently, it spawns all the trains at Station 1, then simulates their path along the track all in one frame, until they reach their actual stopping points. This can cause a noticeable bit of lag when spawning the trains. With this change, it could simply use the list of spawn locations, completely eliminating Ride::MoveTrainsToBlockBrakes(). Maybe there's some reason to do it the current way that I'm unaware of, but if not, it could simplify things quite a bit.

There's also an existing bug that could interfere with this: If you build a station piece in simulate mode, it breaks the station indices. If you're building a ride in a launch mode, this straight up breaks the ride permanently, but it silently occurs with all other rides too. That might interact with this in weird ways.

This is just a rough POC, there's still a lot of work that it would need - I did break block section mode a little, and I'm not sure how to handle it if the overflow from one station intersects with the next.

**Reply by ocalhoun6** [reply]
Maybe there could be a version where this behavior is disabled for rides in block section mode and rides that can have more trains than fit in the station?

Could start with this basic case, and then work on the special cases at another time, gradually working on it until it works well for all multi-station rides.

**Reply by 733737** [reply]
I propose you submit a PR to enhance/improve block brakes first separately

**Comment by HtotheTML** [comment]
Would there be a way to make it a togglable option to check or uncheck in ride operation settings, in the event someone somehow doesn't want that, or would that not be a feasible or good idea?

**Reply by ocalhoun6** [reply]
Assuming it's working properly, without bugs even in edge cases, why would you want to disable it?

In what situation is it *desirable* to spawn all the trains at a single station of a multi-station ride?

**Reply by HtotheTML** [reply]
Can't think of one, honestly.

**Comment by autosysops** [comment]
If this is implemented it should for sure be a toggle as it would else possibly break a lot of "shoestring" rides which often have multiple stations but not all are used the same.

**Comment by Gymnasiast** [comment, marked as answer]
https://github.com/OpenRCT2/OpenRCT2/pull/23638

**Reply by sjoerddebruin** [reply]
Note this pull request wasn't merged but closed so this discussion probably has to be "reopened".
